File: src/main/java/de/schramm/royalbash/controller/service/core/effect/DealDamageToCreatureEffect.java
package de.schramm.royalbash.controller.service.core.effect;

import de.schramm.royalbash.controller.service.core.Context;
import de.schramm.royalbash.controller.service.core.Game;
import lombok.Builder;
import lombok.Value;
import lombok.val;

@Value
@Builder
public class DealDamageToCreatureEffect {

    private final int amountOfDamage;

    Game invoke(final Context context) {

        val game = context.getGame();

        return context.getTargetCreature()
                .map(creature -> game.updateCreature(creature, creature.damage(amountOfDamage)))
                .orElse(game);
    }
}

File: src/main/kotlin/de/schramm/royalbash/controller/service/core/effect/DealDamageToCreatureEffect.kt
package de.schramm.royalbash.controller.service.core.effect

import de.schramm.royalbash.controller.service.core.Context
import de.schramm.royalbash.controller.service.core.Game

data class DealDamageToCreatureEffect(private val amountOfDamage: Int) {

    operator fun invoke(context: Context): Game {

        val game = context.game

        return context.getTargetCreature()
                .map { creature -> game.updateCreature(creature, creature.damage(amountOfDamage)) }
                .orElse(game)
    }
}
